packages/plainjs/dist/deno/optimistic/tokenizer.ts

```diff
@@ -0,0 +1,788 @@
+            if (
+              !this.separatorBytes ||
+              n !== this.separatorBytes[this.separatorIndex]
+            ) {
+              break;
+            }
+            if (this.separatorIndex === this.separatorBytes.length - 1) {
+              this.state = OptimisticTokenizerStates.START;
+              this.onToken({
+                token: TokenType.SEPARATOR,
+                value: this.separator as string,
+                offset: this.offset + this.separatorIndex,
+                type: "complete",
+              });
+              this.separatorIndex = 0;
+            }
+            continue;
+          case OptimisticTokenizerStates.ENDED:
+            if (
+              n === charset.SPACE ||
+              n === charset.NEWLINE ||
+              n === charset.CARRIAGE_RETURN ||
+              n === charset.TAB
+            ) {
+              // whitespace
+              continue;
+            }
+        }
+
+        throw new OptimisticTokenizerError(
+          `Unexpected "${String.fromCharCode(
+            n
+          )}" at position "${i}" in state ${OptimisticTokenizerStateToString(
+            this.state
+          )}`
+        );
+      }
+
+      switch (this.state) {
+        case OptimisticTokenizerStates.TRUE1:
+        case OptimisticTokenizerStates.TRUE2:
+        case OptimisticTokenizerStates.TRUE3:
+          this.onToken({
+            token: TokenType.TRUE,
+            value: true,
+            offset: this.offset,
+            type: "incomplete",
+          });
+          break;
+        case OptimisticTokenizerStates.FALSE1:
+        case OptimisticTokenizerStates.FALSE2:
+        case OptimisticTokenizerStates.FALSE3:
+        case OptimisticTokenizerStates.FALSE4:
+          this.onToken({
+            token: TokenType.FALSE,
+            value: false,
+            offset: this.offset,
+            type: "incomplete",
+          });
+          break;
+        case OptimisticTokenizerStates.NULL1:
+        case OptimisticTokenizerStates.NULL2:
+        case OptimisticTokenizerStates.NULL3:
+          this.onToken({
+            token: TokenType.NULL,
+            value: null,
+            offset: this.offset,
+            type: "incomplete",
+          });
+          break;
+        case OptimisticTokenizerStates.STRING_DEFAULT: {
+          const string = this.bufferedString.toString();
+          this.onToken({
+            token: TokenType.STRING,
+            value: string,
+            offset: this.offset,
+            type: "incomplete",
+          });
+          break;
+        }
+        case OptimisticTokenizerStates.NUMBER_AFTER_INITIAL_ZERO:
+        case OptimisticTokenizerStates.NUMBER_AFTER_INITIAL_NON_ZERO:
+        case OptimisticTokenizerStates.NUMBER_AFTER_DECIMAL:
+        case OptimisticTokenizerStates.NUMBER_AFTER_E_AND_DIGIT:
+          this.onToken({
+            token: TokenType.NUMBER,
+            value: this.parseNumber(this.bufferedNumber.toString()),
```

Comment:
This is the only place where this could fail.

If the number is complex `-2`, `2.3`, `2e10`, etc..., the parsing can fail.
I think that in that case we need to decide if not emit, emit with undefined value, or emit with a string value.

Wdyt?

Reply:
If there's a chance it can fail, 100% agree. That said, is there that chance? In my head:

* `NUMBER_AFTER_INITIAL_ZERO` -- the only buffer we can have here is `"-0"`, I think, which will parse as `0`
* `NUMBER_AFTER_INITIAL_NON_ZERO` -- I think here we have buffers like `-1`, `2`, which again will parse
* `NUMBER_AFTER_DECIMAL` -- `-3.1`, `2.2`, which will parse
* `NUMBER_AFTER_E_AND_DIGIT` -- I think things like `2e4`, which will parse

Wholly possible I'm missing cases that would fail though and it's cheap to add the guard either way. Let me know!